File: backend-go/pkg/database/models/task_run.go
package models

import (
	"time"

	"github.com/google/uuid"
	"gorm.io/datatypes"
)

type TaskRun struct {
	ID                uuid.UUID      `json:"id" gorm:"type:uuid;primaryKey;default:uuid_generate_v4()"`
	Kind              string         `json:"kind" gorm:"type:varchar(50);not null"`
	TaskName          string         `json:"task_name" gorm:"type:varchar(120);not null;index"`
	Status            string         `json:"status" gorm:"type:varchar(50);not null;index"`
	OrganizationID    *uuid.UUID     `json:"organization_id" gorm:"type:uuid;index"`
	UserID            *uuid.UUID     `json:"user_id" gorm:"type:uuid;index"`
	TriggeredByUserID *uuid.UUID     `json:"triggered_by_user_id" gorm:"type:uuid;index"`
	TriggerSource     string         `json:"trigger_source" gorm:"type:varchar(50);not null;default:'manual'"`
	ParentRunID       *uuid.UUID     `json:"parent_run_id" gorm:"type:uuid;index"`
	Summary           *string        `json:"summary" gorm:"type:text"`
	ErrorSummary      *string        `json:"error_summary" gorm:"type:text"`
	InputPayload      datatypes.JSON `json:"input_payload" gorm:"type:jsonb;default:'{}'"`
	OutputSummary     datatypes.JSON `json:"output_summary" gorm:"type:jsonb;default:'{}'"`
	Metrics           datatypes.JSON `json:"metrics" gorm:"type:jsonb;default:'{}'"`
	StartedAt         *time.Time     `json:"started_at"`
	CompletedAt       *time.Time     `json:"completed_at"`
	CreatedAt         time.Time      `json:"created_at" gorm:"autoCreateTime"`
	UpdatedAt         time.Time      `json:"updated_at" gorm:"autoUpdateTime"`
}

func (TaskRun) TableName() string {
	return "task_run"
}

type TaskRunStep struct {
	ID           uuid.UUID      `json:"id" gorm:"type:uuid;primaryKey;default:uuid_generate_v4()"`
	TaskRunID    uuid.UUID      `json:"task_run_id" gorm:"type:uuid;not null;index"`
	StepKey      string         `json:"step_key" gorm:"type:varchar(120);not null"`
	StepName     string         `json:"step_name" gorm:"type:varchar(255);not null"`
	Status       string         `json:"status" gorm:"type:varchar(50);not null;index"`
	Summary      *string        `json:"summary" gorm:"type:text"`
	ErrorSummary *string        `json:"error_summary" gorm:"type:text"`
	Metrics      datatypes.JSON `json:"metrics" gorm:"type:jsonb;default:'{}'"`
	StartedAt    *time.Time     `json:"started_at"`
	CompletedAt  *time.Time     `json:"completed_at"`
	CreatedAt    time.Time      `json:"created_at" gorm:"autoCreateTime"`
	UpdatedAt    time.Time      `json:"updated_at" gorm:"autoUpdateTime"`
}

func (TaskRunStep) TableName() string {
	return "task_run_step"
}

type TaskRunEvent struct {
	ID            uuid.UUID      `json:"id" gorm:"type:uuid;primaryKey;default:uuid_generate_v4()"`
	TaskRunID     uuid.UUID      `json:"task_run_id" gorm:"type:uuid;not null;index"`
	TaskRunStepID *uuid.UUID     `json:"task_run_step_id" gorm:"type:uuid;index"`
	EventType     string         `json:"event_type" gorm:"type:varchar(120);not null"`
	Level         string         `json:"level" gorm:"type:varchar(20);not null;default:'info'"`
	Message       string         `json:"message" gorm:"type:text;not null"`
	MetaData      datatypes.JSON `json:"meta_data" gorm:"type:jsonb;default:'{}'"`
	CreatedAt     time.Time      `json:"created_at" gorm:"autoCreateTime"`
}

func (TaskRunEvent) TableName() string {
	return "task_run_event"
}

File: backend/src/database/models/task_run.rs
use chrono::{DateTime, Utc};
use diesel::{AsChangeset, Identifiable, Insertable, Queryable, Selectable};
use serde_json::Value;
use uuid::Uuid;

use crate::schema::{task_run, task_run_event, task_run_step};

#[derive(Debug, Clone, Queryable, Selectable, Identifiable)]
#[diesel(table_name = task_run)]
#[diesel(check_for_backend(diesel::pg::Pg))]
pub struct TaskRunRow {
    pub id: Uuid,
    pub kind: String,
    pub task_name: String,
    pub status: String,
    pub organization_id: Option<Uuid>,
    pub user_id: Option<Uuid>,
    pub triggered_by_user_id: Option<Uuid>,
    pub trigger_source: String,
    pub parent_run_id: Option<Uuid>,
    pub summary: Option<String>,
    pub error_summary: Option<String>,
    pub input_payload: Option<Value>,
    pub output_summary: Option<Value>,
    pub metrics: Option<Value>,
    pub started_at: Option<DateTime<Utc>>,
    pub completed_at: Option<DateTime<Utc>>,
    pub created_at: Option<DateTime<Utc>>,
    pub updated_at: Option<DateTime<Utc>>,
}

#[derive(Debug, Clone, Insertable)]
#[diesel(table_name = task_run)]
pub struct NewTaskRunRow {
    pub id: Uuid,
    pub kind: String,
    pub task_name: String,
    pub status: String,
    pub organization_id: Option<Uuid>,
    pub user_id: Option<Uuid>,
    pub triggered_by_user_id: Option<Uuid>,
    pub trigger_source: String,
    pub parent_run_id: Option<Uuid>,
    pub summary: Option<String>,
    pub error_summary: Option<String>,
    pub input_payload: Value,
    pub output_summary: Value,
    pub metrics: Value,
    pub started_at: Option<DateTime<Utc>>,
    pub completed_at: Option<DateTime<Utc>>,
}

#[derive(Debug, Clone, AsChangeset)]
#[diesel(table_name = task_run)]
pub struct TaskRunChangeset {
    pub kind: String,
    pub task_name: String,
    pub status: String,
    pub organization_id: Option<Option<Uuid>>,
    pub user_id: Option<Option<Uuid>>,
    pub triggered_by_user_id: Option<Option<Uuid>>,
    pub trigger_source: String,
    pub parent_run_id: Option<Option<Uuid>>,
    pub summary: Option<Option<String>>,
    pub error_summary: Option<Option<String>>,
    pub input_payload: Option<Option<Value>>,
    pub output_summary: Option<Option<Value>>,
    pub metrics: Option<Option<Value>>,
    pub started_at: Option<Option<DateTime<Utc>>>,
    pub completed_at: Option<Option<DateTime<Utc>>>,
    pub created_at: Option<Option<DateTime<Utc>>>,
    pub updated_at: Option<DateTime<Utc>>,
}

#[derive(Debug, Clone, Queryable, Selectable, Identifiable)]
#[diesel(table_name = task_run_step)]
#[diesel(check_for_backend(diesel::pg::Pg))]
pub struct TaskRunStepRow {
    pub id: Uuid,
    pub task_run_id: Uuid,
    pub step_key: String,
    pub step_name: String,
    pub status: String,
    pub summary: Option<String>,
    pub error_summary: Option<String>,
    pub metrics: Option<Value>,
    pub started_at: Option<DateTime<Utc>>,
    pub completed_at: Option<DateTime<Utc>>,
    pub created_at: Option<DateTime<Utc>>,
    pub updated_at: Option<DateTime<Utc>>,
}

#[derive(Debug, Clone, Insertable)]
#[diesel(table_name = task_run_step)]
pub struct NewTaskRunStepRow {
    pub id: Uuid,
    pub task_run_id: Uuid,
    pub step_key: String,
    pub step_name: String,
    pub status: String,
    pub summary: Option<String>,
    pub error_summary: Option<String>,
    pub metrics: Value,
    pub started_at: Option<DateTime<Utc>>,
    pub completed_at: Option<DateTime<Utc>>,
}

#[derive(Debug, Clone, AsChangeset)]
#[diesel(table_name = task_run_step)]
pub struct TaskRunStepChangeset {
    pub step_key: String,
    pub step_name: String,
    pub status: String,
    pub summary: Option<Option<String>>,
    pub error_summary: Option<Option<String>>,
    pub metrics: Option<Option<Value>>,
    pub started_at: Option<Option<DateTime<Utc>>>,
    pub completed_at: Option<Option<DateTime<Utc>>>,
    pub created_at: Option<Option<DateTime<Utc>>>,
    pub updated_at: Option<DateTime<Utc>>,
}

#[derive(Debug, Clone, Queryable, Selectable, Identifiable)]
#[diesel(table_name = task_run_event)]
#[diesel(check_for_backend(diesel::pg::Pg))]
pub struct TaskRunEventRow {
    pub id: Uuid,
    pub task_run_id: Uuid,
    pub task_run_step_id: Option<Uuid>,
    pub event_type: String,
    pub level: String,
    pub message: String,
    pub meta_data: Option<Value>,
    pub created_at: Option<DateTime<Utc>>,
}

#[derive(Debug, Clone, Insertable)]
#[diesel(table_name = task_run_event)]
pub struct NewTaskRunEventRow {
    pub id: Uuid,
    pub task_run_id: Uuid,
    pub task_run_step_id: Option<Uuid>,
    pub event_type: String,
    pub level: String,
    pub message: String,
    pub meta_data: Value,
}
